crates/shared/src/trade_finding/oneinch.rs

```diff
@@ -0,0 +1,298 @@
+//! A 1Inch-based trade finder.
+
+use super::{Interaction, Query, Trade, TradeError, TradeFinding};
+use crate::{
+    oneinch_api::{OneInchClient, OneInchError, ProtocolCache, SellOrderQuoteQuery, SwapQuery},
+    price_estimation::gas,
+    solver_utils::Slippage,
+};
+use model::order::OrderKind;
+use primitive_types::H160;
+use std::sync::Arc;
+
+pub struct OneInchTradeFinder {
+    api: Arc<dyn OneInchClient>,
+    disabled_protocols: Vec<String>,
+    protocol_cache: ProtocolCache,
+    referrer_address: Option<H160>,
+}
+
+impl OneInchTradeFinder {
+    async fn quote_and_swap(&self, query: &Query) -> Result<Trade, TradeError> {
+        if query.kind == OrderKind::Buy {
+            return Err(TradeError::UnsupportedOrderType);
+        }
+
+        let allowed_protocols = self
+            .protocol_cache
+            .get_allowed_protocols(&self.disabled_protocols, self.api.as_ref())
+            .await?;
+
+        let (quote, spender, swap) = futures::try_join!(
+            self.api
+                .get_sell_order_quote(SellOrderQuoteQuery::with_default_options(
+                    query.sell_token,
+                    query.buy_token,
+                    allowed_protocols.clone(),
+                    query.in_amount,
+                    self.referrer_address,
+                )),
+            self.api.get_spender(),
```

Comment:
Should we maybe cache the `/spender` call a bit?

Reply:
I would hope that they do that on their end. In my tests it returns in ~130ms so I don't think it will be a limiting factor:
```
% /usr/bin/time -h bash -c 'for i in $(seq 0 100); do curl -s 'https://api.1inch.io/v4.0/1/approve/spender' >/dev/null; done'
	13.07s real	1.62s user		1.00s sys
```

Will create an issue to cache this though.